card/lib/card/view_name.rb

```diff
@@ -0,0 +1,44 @@
+# -*- encoding : utf-8 -*-
+require_dependency 'card/env'
+
+require 'smart_name'
+
+class Card
+  class ViewName < SmartName
+    @@name2viewnameobject = {}
+
+    class << self
+      def new obj
+        return obj if self.class===obj
+        str = Array===obj ? obj*joint : obj.to_s
+        if known_name = @@name2viewnameobject[str]
+          known_name
+        else
+          super str.strip
+        end
+      end
+    end
+
+    def initialize str
+      @s = str.to_s.strip
+      @s = @s.encode('UTF-8') if RUBYENCODING
+      @key = if @s.index(self.class.joint)
+          @parts = @s.split(/\s*#{JOINT_RE}\s*/)
+          @parts << '' if @s[-1,1] == self.class.joint
+          @simple = false
+          @parts.map { |p| p.to_name.key } * self.class.joint
+        else
+          @parts = [str]
+          @simple = true
+          str.empty? ? '' : simple_key
+        end
+      @@name2viewnameobject[str] = self
+    end
+
+    def simple_key
+      decoded.underscore.gsub(/[^#{OK4KEY_RE}]+/,'_').split(/_+/).reject(&:empty?)*'_'
+    end
```

Comment:
It would be better if we didn't have to clone some of the code in smartname.  Is there a clean way to make that processing optional?  In the end, I think it will be really hard to do something like this without it being part of the way names are handled generally.  What I mean is that if we ever will have the view names actually represented by cards, and it seems that we will,, then we can't use an option to do view name translation to codenames that isn't in effect whenever we are doing name lookup or we won't be able to find cards with plural parts to their keys.
If we process view names the same as cardnames are processed now, what is the downside?  I think it is just that we sometimes have odd looking codenames for some view, right?  That's a lot easier to live with than inconsistent handling of name to key translations.